airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/StateMessageHelper.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.protocol.models;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.commons.json.Jsons;
+import io.vavr.control.Either;
+import java.util.ArrayList;
+import java.util.List;
+
+public class StateMessageHelper {
+
+  public static class AirbyteStateMessageListTypeReference extends TypeReference<List<AirbyteStateMessage>> {}
+
+  /**
+   * This a takes a json blob state and tries return either a legacy state in the format of a json
+   * object or a state message with the new format which is a list of airbyte state message.
+   *
+   * @param state
+   * @return Either a json blob (on the left) or a structure state message.
+   */
+  public static Either<JsonNode, List<AirbyteStateMessage>> getTypedState(JsonNode state) {
```

Comment:
what's the use case for this? not obvious to me why we wouldn't always return a `List<AirbyteStateMessage>`

Reply:
It is to make it extra clear that there is a split between legacy and new message. In legacy we don't have a typed message, the goal here is to be as close as possible to the format of the state: https://github.com/airbytehq/airbyte/pull/13772#discussion_r897356112

Reply:
https://github.com/spencerwi/Either.java the first part of the description looks like a similar use case to what we are trying to achieve here which is to let the user of this library that they have to deal with a state in the old format (just a blob) or a state in the new format (List of messages).

Reply:
Would it make sense to use the `StateWrapper` that you mentioned in the [spec](https://docs.google.com/document/d/10cpbytxVzuwSpLtAA4fQeDoXRE17sY81h0bTrUPSE7s/edit#heading=h.7gaa7py6gxwq)?

Reply:
It seems like `Either` is only doing part of the job. In your calling code in the other PR, you are still doing a ton of work to try to figure out what the type of the state message is. I think you want something more descriptive than `Either`. What you mentioned in the spec seems like it would be awesome here. And definitely have a static helper to construct that object from json would be useful!

Reply:
I really wanted to be as close as possible to how we represent the state in the JsonBlob which is either a blob or a list of state message. Then understanding what is the type of the state would be done somewhere else. 

I'll modify it to return the wrapper. @gosusnp Have you introduce the class in your PR?

Reply:
why?

Reply:
Yes, I am adding the `StateWrapper` in my PR. I'll share a draft once I updated to the latest version.

Reply:
I agree with Charles here - rather than having this function turn a JsonNode state into an intermediate `Either` object and then requiring consumers of this to interpret that into the state type, it seems better to make this the one place that knows how to determine the state type completely from the JsonNode. So I think `StateWrapper` would be great here

Reply:
I will update it. I am working with @gosusnp to get it in a PR.